Store classes not worth remembering per client for JITServer

dmitry-ten opened the pull request:
PR #10644 added a static array of classes not worth remembering
to Symbol Validation Manager. JITServer, however, requires this array
to be stored on a per client bases, since pointers to J9 classes
are different between clients.

dmitry-ten commented:
@mpirvu this should fix #10663. I started the tests, will update when they complete.

mpirvu commented on runtime/compiler/runtime/SymbolValidationManager.cpp:
```diff
@@ -111,6 +115,26 @@ TR::SymbolValidationManager::SymbolValidationManager(TR::Region &region, TR_Reso
+
+#if defined(J9VM_OPT_JITSERVER)
+   if (TR::CompilationInfo::get()->getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
+      {
+      // If needed, populate the client session with classes not worth remembering
+      auto clientData = _fej9->_compInfoPT->getClientData();
+      auto &classesNotWorthRemembering = clientData->getSystemClassesNotWorthRemembering();
+         {
+         OMR::CriticalSection classesNotWorthRememberingMonitor(clientData->getClassesNotWorthRememberingMonitor());
+         if (classesNotWorthRemembering.size() != SYSTEM_CLASSES_NOT_WORTH_REMEMBERING_COUNT)
+            {
+            for (int i = 0; i < SYSTEM_CLASSES_NOT_WORTH_REMEMBERING_COUNT; ++i)
+               {
+               auto classNotWorthRemembering = &_systemClassesNotWorthRemembering[i];
+               classesNotWorthRemembering.push_back({ classNotWorthRemembering->_className, NULL, classNotWorthRemembering->_checkIsSuperClass });
+               }
+            }
+         }
+      }
```
I think we should execute this block of code in the constructor of a session where our new vector of classes not worth remembering is defined. This way we can avoid locking getClassesNotWorthRememberingMonitor() every time we create an SVM object. I am proposing to take this block of code (more or less) and create a static function in `SymbolValidationManager`. Then call this function in the constructor of a client session. We may not even need a vector, just an array of structs should be enough, though I don't know if we can use the computed number of elements in another file.

dmitry-ten replied:
Yeah, I will implement that. I think we don't even need a monitor if array initialization is done in client session constructor.

mpirvu reviewed (commented):
It looks mostly ok. Please see the inline suggestions

mpirvu commented:
Not sure if this can be used to determine the number of elements in the array, since the array is defined in cpp while we need this definition in hpp:
```

dmitry-ten commented on runtime/compiler/runtime/SymbolValidationManager.cpp:
```diff
@@ -111,6 +115,26 @@ TR::SymbolValidationManager::SymbolValidationManager(TR::Region &region, TR_Reso
       }
+
+#if defined(J9VM_OPT_JITSERVER)
+   if (TR::CompilationInfo::get()->getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
```
We can't use that method here, since SVM is created from `J9::Compilation` constructor, and `_isOutOfProcessCompilation` is not set yet. But if we move if outside of the constructor then it can be changed.

dmitry-ten commented on runtime/compiler/control/JITClientCompilationThread.cpp:
```diff
@@ -256,9 +256,10 @@ handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe, JITServer::Mes
       case MessageType::VM_getSystemClassFromClassName:
          {
          auto recv = client->getRecvData<std::string, bool>();
+         TR_J9VMBase *fej9 = TR_J9VMBase::get(vmThread->javaVM->jitConfig, vmThread);
```
Yeah, we need non-AOT front-end here, since validation is done on the server. I mention it in the commit message, will add comment here as well.

mpirvu replied:
Ironically, I just reviewed an item from @dchopra001 where I suggested using a TR_J9 frontend in this very place. In a sense I answered my own question in that PR.

dchopra001 replied:
Can we remove this line of code from here and merge it as part of #10688 instead? The change here is needed to functional correctness on Z. Since we are trying to get this change into the 0.23 branch, it'll be easier to push it through as part of #10688. Unless we are sure that the change in this PR should also be double delivered?

dmitry-ten replied:
I will remove it.

dchopra001 replied:
Thanks!

dmitry-ten commented:
@mpirvu other than replacing vector with array, I addressed all comments.
Replacing vector with array requires creating a dynamically allocated array though, which needs default constructor. And to assign to that array, assignment operator is also needed. I'm not sure if the additional code would be worth it.
@dsouzai could you also take a look?

mpirvu reviewed (approved):
LGTM

mpirvu commented:
We could use an embedded array if we knew the size. If the size of the original array cannot be statically determined, an workaround is to declare it as '2' in the hpp file and then to add a check at compile time that the number of entries in the cpp is actually 2. I'll let @dsouzai decide if he is ok with that change

dsouzai commented:
Yeah looks ok to me.

> If the size of the original array cannot be statically determined, an workaround is to declare it as '2' in the hpp file and then to add a check at compile time that the number of entries in the cpp is actually 2.

Yeah I'm ok with this since it'll be a compile time static assert, so it will at least get caught pretty early on.

dmitry-ten commented:
@mpirvu changed vector to static array and removed the fix to front-end for `getSystemClassFromSignature` message.

mpirvu commented on runtime/compiler/runtime/SymbolValidationManager.hpp:
```diff
@@ -637,9 +641,9 @@ class SymbolValidationManager
    struct SystemClassNotWorthRemembering
       {
-      const char * const _className;
+      char *_className;
```
Can't we leave the first `const` here. The second one refers to the `_className` being immutable, but the first one refers to the string of characters being immutable.

dmitry-ten replied:
True. I can push the change once the tests pass.

mpirvu replied:
The tests on Power take very long for some reason, but on x86 they passed. You could do this change now.

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit jdk11

dmitry-ten commented:
Pushed

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit jdk11

mpirvu commented:
Testing on x86 was successful. Testing on Power seems to hand for some reason even though all the tests have passed. I don't know what the infra is waiting for.
I am going to merge this item since it addresses some serious breakage for JITServer.
